cgp/genome.py

```diff
@@ -267,6 +267,172 @@ def randomize(self, rng: np.random.RandomState) -> None:
+            region_idx = gene_idx // length_per_region
+            if self._is_hidden_input_gene(gene_idx, region_idx) or self._is_output_input_gene(
+                gene_idx
+            ):
+                try:
+                    gene_value = self._n_inputs + used_idx_list.index(gene_value)
+                except ValueError:  # means the input is from an input node -> do nothing
+                    pass
+            dna_for_reorder[gene_idx] = gene_value
+
+        # change the invalid gene values (coming from unused arities)
+        for gene_idx, gene_value in enumerate(dna_for_reorder):
+            region_idx = gene_idx // length_per_region
+            if self._is_hidden_input_gene(gene_idx, region_idx) and gene_value > region_idx:
+                # replace value that is too large by smaller value
+                permissible_values = self.determine_permissible_values_per_gene(gene_idx)
+                gene_value = rng.choice(permissible_values)
+                dna_for_reorder[gene_idx] = gene_value
+
+        self.dna = dna_for_reorder
+
+    def _determine_dependencies(self) -> List[set]:
+
+        dependencies = []
+        for region_idx in range(self._n_regions):
+            current_node_dependencies = set()
+
+            if self._is_input_region(region_idx):
+                dependencies.append(
+                    current_node_dependencies
+                )  # input regions have no dependencies
+
+            elif self._is_hidden_region(
+                region_idx
+            ):  # can assume dependencies is not empty anymore
+                # first determine arity of current operator
+                # to only consider genes which are used in current node
+                operator_idx = region_idx * self._length_per_region
+                operator_value = self._dna[operator_idx]
+                current_arity = self._primitives[operator_value]._arity
+
+                # append each input and its respective dependencies to the dependencies
+                for idx_gene in range(
+                    1, current_arity + 1
+                ):  # shift by 1 since first gene is operator
+                    input_region_idx = self._dna[operator_idx + idx_gene]
+                    if not self._is_input_region(
+                        input_region_idx
+                    ):  # not necessary to add input regions, since they are first anyway
+                        current_node_dependencies.add(input_region_idx)  # append the input node &
+                        current_node_dependencies.update(
+                            dependencies[input_region_idx]
+                        )  # the dependencies of the input node to the current node dependencies
+                        # Todo: can be removed?
+                dependencies.append(current_node_dependencies)
+
+            elif self._is_output_region(
+                region_idx
+            ):  # Todo: actually it is not necessary to calculate those,
```

Comment:
judging by your comment, a `continue` would also do here? ;) if so, you could combine it with the `continue` for input regions